### PR TITLE
[module-minifier-plugin] Upgrade terser to 5.10.0

### PR DESCRIPTION
**Pull request by dmichon-msft**
## Summary
Upgrades ModuleMinifierPlugin to use terser 5.10.0, removing the need to monkey patch identifier allocation.

## Details
Uses the async minifier API on terser and takes advantage of the new `mangle.nth_identifier` hook to customize identifier allocation and ensure uniformity across minified modules.
Updates NoopMinifier to emit an empty source map if source maps are requested.

## How it was tested
Using the localization-plugin-02 build test.

### CODE REVIEW

File: common/reviews/api/module-minifier-plugin.api.md [hunk 1]
```diff
@@ -205,14 +215,6 @@ export const STAGE_AFTER: 100;
 // @public
 export const STAGE_BEFORE: -100;
 
-// @public
-export class SynchronousMinifier implements IModuleMinifier {
-    constructor(options: ISynchronousMinifierOptions);
-    minify(request: IModuleMinificationRequest, callback: IModuleMinificationCallback): void;
-    // (undocumented)
-    readonly terserOptions: MinifyOptions;
```
Comment (D4N14L) [inline common/reviews/api/module-minifier-plugin.api.md:213]: Should probably be documented

File: webpack/module-minifier-plugin/src/NoopMinifier.ts [hunk 2]
```diff
@@ -18,13 +18,23 @@ export class NoopMinifier implements IModuleMinifier {
...
-      map: undefined
+      // If source maps are requested, provide an empty source map
+      map: nameForMap
+        ? {
+            version: 3,
+            names: [],
+            file: nameForMap,
+            sources: [nameForMap],
+            sourcesContent: [code],
+            mappings: 'AAAA'
```
Comment (D4N14L) [inline webpack/module-minifier-plugin/src/NoopMinifier.ts:35]: This seems like an odd default for `mappings

File: webpack/module-minifier-plugin/src/LocalMinifier.ts [hunk 3]
```diff
@@ -56,9 +56,24 @@ export class SynchronousMinifier implements IModuleMinifier {
       return callback(cached);
     }
 
-    const result: IModuleMinificationResult = minifySingleFile(request, this.terserOptions);
-    this._resultCache.set(hash, result);
-
-    callback(result);
+    minifySingleFile(request, this.terserOptions)
+      .then((result: IModuleMinificationResult) => {
+        this._resultCache.set(hash, result);
```
Comment (D4N14L) [inline webpack/module-minifier-plugin/src/LocalMinifier.ts:61]: Looking at `minifySingleFile` it seems like it's designed already to catch errors and re-output like below. Would we want to cache if there's an error present in the `result`?

File: webpack/module-minifier-plugin/src/LocalMinifier.ts [hunk 4]
```diff
@@ -56,9 +56,24 @@ export class SynchronousMinifier implements IModuleMinifier {
...
+        });
+      });
   }
 }
+
+/**
+ * Use the name "LocalMinifier" instead for technical correctness.
+ * @deprecated
+ */
+export { LocalMinifier as SynchronousMinifier };
```
Comment (D4N14L) [inline webpack/module-minifier-plugin/src/LocalMinifier.ts:79]: This seems odd... especially since `minify` isn't synchronous anymore.

**Review by D4N14L — APPROVED** [review]

File: common/reviews/api/module-minifier-plugin.api.md [hunk 5]
```diff
@@ -205,14 +211,6 @@ export const STAGE_AFTER: 100;
 // @public
 export const STAGE_BEFORE: -100;
 
-// @public
-export class SynchronousMinifier implements IModuleMinifier {
```
Comment (iclanton) [inline common/reviews/api/module-minifier-plugin.api.md:209]: This is a breaking change.

Comment (D4N14L) [reply common/reviews/api/module-minifier-plugin.api.md:209]: David and I discussed this earlier, but I think a breaking change is fair here. For one, it no longer runs synchronously so labeling it as such is a bit misleading. Also, it's still a `0.x.x` package so the minor bump behaves like a major bump in terms of `~` and `^` versions in `package.json`.

Comment (iclanton) [reply common/reviews/api/module-minifier-plugin.api.md:209]: Sounds reasonable, then.

**Review by iclanton — APPROVED** [review]

**Pull request merged**